### PR TITLE
Check if required setters are set

### PR DESCRIPTION
**Pull request by phanimarupaka**
@mortent @pwittrock 

This PR is to make check for all the setters and return error if a package consumer didn't set the setters which are marked as required by package publisher. This is a follow up PR for the PR in kustomize kubernetes-sigs/kustomize#2635. Description in the kustomize PR and test cases in current PR clearly describe the expected behavior.

The motivation for this PR is to address the feature request https://github.com/GoogleContainerTools/kpt/issues/261

### CODE REVIEW

File: kyaml/setters2/list.go [hunk 1]
```diff
@@ -192,3 +193,44 @@ func (l *List) count(path, name string) (int, error) {
...
+
+	for _, key := range keys {
+		required, err := definitions.Pipe(yaml.Lookup(key, "x-k8s-cli", "setter", "required"))
+		if required == nil || err != nil {
+			continue
+		}
+		requiredVal := required.Document().Value
+		println(requiredVal)
+
+		isSet, err := definitions.Pipe(yaml.Lookup(key, "x-k8s-cli", "setter", "isSet"))
```
Comment (monopole) [inline kyaml/setters2/list.go:228]: maybe just test err != nil by itself?

then check the requiredVal stuff?

File: kyaml/setters2/list.go [hunk 2]
```diff
@@ -192,3 +193,44 @@ func (l *List) count(path, name string) (int, error) {
...
+		return nil
+	}
+
+	for _, key := range keys {
+		required, err := definitions.Pipe(yaml.Lookup(key, "x-k8s-cli", "setter", "required"))
+		if required == nil || err != nil {
+			continue
+		}
+		requiredVal := required.Document().Value
+		println(requiredVal)
```
Comment (monopole) [inline kyaml/setters2/list.go:226]: is println intentional?

**Comment by mortent** [comment]
As discussed offline, I think we should try to check this using the openapi api, assuming any krm files have already been read and any definitions merged into the schema.

File: kyaml/setters2/list.go (hunk 2 above)
Comment (phanimarupaka) [reply kyaml/setters2/list.go]: Seeped in by mistake. Thanks.

File: kyaml/setters2/list.go (hunk 1 above)
Comment (phanimarupaka) [reply kyaml/setters2/list.go]: Updated the PR with new implementation.

**Comment by phanimarupaka** [comment]
@mortent Done.

File: kyaml/setters2/list.go [hunk 3]
```diff
@@ -192,3 +192,22 @@ func (l *List) count(path, name string) (int, error) {
 
 	return s.Count, err
 }
+
+// CheckRequiredSettersSet iterates through all the setter definitions in openAPI
+// schema and returns error if any of the setter has required field true and isSet false
+func CheckRequiredSettersSet() error {
```
Comment (mortent) [inline kyaml/setters2/list.go:198]: nit: Could we put this at least in a separate file? It doesn't seem to really have anything to do with listing setters.

**Comment by mortent** [comment]
Looks good to me. I'll give @pwittrock some time to take a look too.

File: kyaml/setters2/list.go (hunk 3 above)
Comment (pwittrock) [reply kyaml/setters2/list.go:198]: +1

File: kyaml/setters2/list_test.go [hunk 4]
```diff
@@ -287,3 +289,160 @@ spec:
...
+	}
+	for i := range tests {
+		test := tests[i]
+		t.Run(test.name, func(t *testing.T) {
+			openapi.ResetOpenAPI()
+			defer openapi.ResetOpenAPI()
+			dir, err := ioutil.TempDir("", "")
+			assert.NoError(t, err)
+			defer os.RemoveAll(dir)
+			err = ioutil.WriteFile(filepath.Join(dir, "kustomization"), []byte(test.inputOpenAPIfile), 0600)
```
Comment (pwittrock) [inline kyaml/setters2/list_test.go:429]: Use `Krmfile` instead

**Comment by pwittrock** [comment]
Looks good from my end.  Address the nits.

File: kyaml/setters2/list_test.go (hunk 4 above)
Comment (phanimarupaka) [reply kyaml/setters2/list_test.go]: Done

File: kyaml/setters2/list.go (hunk 3 above)
Comment (phanimarupaka) [reply kyaml/setters2/list.go]: @mortent Done.

**Pull request merged**